### ISSUE TITLE
Difference between origin EDM traning process and diffusers EDM training process

### ISSUE THREAD

**Issue by bendanzzc** [huggingface/diffusers#6041]
in the original EDM training the σ is sampled from a certain Gaussian distribution
[https://github.com/NVlabs/edm/blob/main/training/loss.py#L73C1-L73C9](https://github.com/NVlabs/edm/blob/main/training/loss.py#L73C1-L73C9)
![image](https://github.com/huggingface/diffusers/assets/26294676/85f44a73-329f-4267-97d0-940d3f0521c8)

in diffuers the σ is sampled from timesteps which I think is precomputed[https://github.com/huggingface/diffusers/blob/main/examples/research_projects/controlnet/train_controlnet_webdataset.py#L1342](https://github.com/huggingface/diffusers/blob/main/examples/research_projects/controlnet/train_controlnet_webdataset.py#L1342)
![image](https://github.com/huggingface/diffusers/assets/26294676/83fa69b2-e5de-4432-b2fb-9a9cfd4d54aa)

could the different training process behave the same?

**Comment by dg845** [comment]
In the [EDM training code](https://github.com/NVlabs/edm/blob/62072d2612c7da05165d6233d13d17d71f213fee/training/loss.py#L72-L74), the lines 

```python
class EDMLoss:
    ...
    def __call__(self, net, images, labels=None, augment_pipe=None):
        rnd_normal = torch.randn([images.shape[0], 1, 1, 1], device=images.device)
        sigma = (rnd_normal * self.P_std + self.P_mean).exp()
        ...
    ...
```

correspond to the training noise distribution in the "EDM" column of Table 1 in the [EDM paper](https://arxiv.org/pdf/2206.00364.pdf):

```math
\ln(\sigma) \sim \mathcal{N}(P_{mean}, P_{std}^2)
```

By contrast, the `EulerDiscreteScheduler` used in the script allows $`\sigma`$ to only take on the discrete values

```math
\Bigl\{ \sqrt{\frac{1 - \bar{\alpha}_t}{\bar{\alpha}_t}} \Bigr\}_{t = 1}^{T}
```

determined by the beta schedule $`\beta_t = 1 - \alpha_t`$:

https://github.com/huggingface/diffusers/blob/4684ea2fe8d568f44c491068c3eb94aac27045f3/src/diffusers/schedulers/scheduling_euler_discrete.py#L166-L170

which are sampled from uniformly at random during training:

https://github.com/huggingface/diffusers/blob/4684ea2fe8d568f44c491068c3eb94aac27045f3/examples/research_projects/controlnet/train_controlnet_webdataset.py#L1335-L1342

In general these approaches are not the same: that is, the `diffusers` ControlNet training script does not follow the EDM training noise distribution design choices.

> could the different training process behave the same?

I haven't used the training script in question but it could very well be the case that both approaches can result in a model that generates good samples. I think it's less likely that both approaches will result in models that contain very similar weights.

**Comment by bendanzzc** [comment]
Thanks for detailed replay

**Comment by bendanzzc** [comment]
> In the [EDM training code](https://github.com/NVlabs/edm/blob/62072d2612c7da05165d6233d13d17d71f213fee/training/loss.py#L72-L74), the lines
> 
> ```python
> class EDMLoss:
>     ...
>     def __call__(self, net, images, labels=None, augment_pipe=None):
>         rnd_normal = torch.randn([images.shape[0], 1, 1, 1], device=images.device)
>         sigma = (rnd_normal * self.P_std + self.P_mean).exp()
>         ...
>     ...
> ```
> 
> correspond to the training noise distribution in the "EDM" column of Table 1 in the [EDM paper](https://arxiv.org/pdf/2206.00364.pdf):
> 
> ln
> ⁡
> (
> �
> )
> ∼
> �
> (
> �
> �
> �
> �
> �
> ,
> �
> �
> �
> �
> 2
> )
> By contrast, the `EulerDiscreteScheduler` used in the script allows σ to only take on the discrete values
> 
>  
> {
> 1
> −
> �
> ¯
> �
> �
> ¯
> �
> }
> �
> =
> 1
> �
> determined by the beta schedule βt=1−αt:
> 
> https://github.com/huggingface/diffusers/blob/4684ea2fe8d568f44c491068c3eb94aac27045f3/src/diffusers/schedulers/scheduling_euler_discrete.py#L166-L170
> 
> which are sampled from uniformly at random during training:
> 
> https://github.com/huggingface/diffusers/blob/4684ea2fe8d568f44c491068c3eb94aac27045f3/examples/research_projects/controlnet/train_controlnet_webdataset.py#L1335-L1342
> 
> In general these approaches are not the same: that is, the `diffusers` ControlNet training script does not follow the EDM training noise distribution design choices. (I'm not sure the `diffusers` implementation _intended_ to use the EDM noise distribution; the [ControlNet paper](https://openaccess.thecvf.com/content/ICCV2023/papers/Zhang_Adding_Conditional_Control_to_Text-to-Image_Diffusion_Models_ICCV_2023_paper.pdf) and [supplementary materials](https://openaccess.thecvf.com/content/ICCV2023/supplemental/Zhang_Adding_Conditional_Control_ICCV_2023_supplemental.pdf) don't mention this as far as I know.)
> 
> > could the different training process behave the same?
> 
> I haven't used the training script in question but it could very well be the case that both approaches can result in a model that generates good samples. I think it's less likely that both approaches will result in models that contain very similar weights. The training behavior for each approach will probably be different: for example, the time to convergence may be different for each approach.

Another question，may be you can give me some suggestions:

**Comment by bendanzzc** [comment]
Another quesiton, may be you can give me some suggestion.

is it possible or reasonable that I mix the DDPM and EDM training scheduler? for example I pretrain the model using EDM training scheduler but finetune the model using DDPM scheduler?

thanks a lot
@dg845
